Add reconcile details (#2612)
operator-framework/operator-sdk: doc/user-guide.md

**Note:** Returning `Result` with `RequeueAfter` set is how you can periodically reconcile a CR.

#### Reconcile Result Use Cases
**The following are possible reconcile loop return options.**

#### 1. With the error:

If an error is encountered during processing the appropriate return option is to return an error.
This results in the reconcile loop being re-triggered to run again.

> Usage

Reviewer: ```suggestion
**Usage**
```

Author: Fixed

```Go
return reconcile.Result{}, err
```

> Example:

Reviewer: ```suggestion
**Example**
```

Author: Fixed

> In the [`memcached_controller.go`][memcached_controller] example below a `reconcile.Result{}, err` is used when there is an error reading the object.

Reviewer: ```suggestion
In the example below a `reconcile.Result{}, err` is used when there is an error reading the object.
```
IHMO  remove the `[`memcached_controller.go`][memcached_controller] ` and would be better in order to keep it maintained and try to be less verbose.

Author: Sure. Fixed.

As a result the request is requeued for another try.
```Go
// Fetch the Memcached instance
memcached := &cachev1alpha1.Memcached{}
err := r.client.Get(context.TODO(), request.NamespacedName, memcached)
if err != nil {
	if errors.IsNotFound(err) {
		...
  }
  // Error reading the object - requeue the request.
	reqLogger.Error(err, "Failed to get Memcached")
	return reconcile.Result{}, err

> }

Reviewer: Indentation of the code above is not accurate.

Author: Good catch! Fixed!

Author: Good catch! Fixed.

```

#### 2. Without an error:

There are several situations where although no error occured, the reconcile loop should signify
during its return that it needs to run again.

> Usage

Reviewer: ```suggestion
**Usage**
```

Author: Fixed

```Go
return reconcile.Result{Requeue: true}, nil
```

> Example:

Reviewer: ```suggestion
**Example**
```

Author: Fixed

> In the [`memcached_controller.go`][memcached_controller] example below a `reconcile.Result{Requeue: true}, nil` is used because a new resources is being created and as such there is the 

Reviewer: ```suggestion
In the example below a `reconcile.Result{Requeue: true}, nil` is used because a new resource is being created and as such, there is the 
```

Reviewer: Check the line is broken when show that should not be

Author: Fixed

potential that further processing is required. 
Thus, the reconcile loop needs to trigger a requeue but there is no error associated with this requeue.
As a result the request is requeued for another try.
```Go
// Define a new deployment
dep := r.deploymentForMemcached(memcached)
...

// Deployment created successfully - return and requeue
return reconcile.Result{Requeue: true}, nil
```

#### 3. Without an error and no need to requeue the request:

In some situations, such as when the primary resource has been deleted, there is no need to
requeue the request for another attempt

> Usage

Reviewer: ```suggestion
**Usage**
```

Author: Fixed

```Go
return reconcile.Result{}, nil
```

> Example:

Reviewer: ```suggestion
**Example**
```

Author: Fixed

> In the [`memcached_controller.go`][memcached_controller] example below a `reconcile.Result{}, nil` is used because the Memcached resource was not found, and no further processing is required.  

Reviewer: ```suggestion
In the example below a `reconcile.Result{}, nil` is used because the Memcached resource was not found, and no further processing is required.  
```

Author: Fixed

```Go
// Fetch the Memcached instance
memcached := &cachev1alpha1.Memcached{}
err := r.client.Get(context.TODO(), request.NamespacedName, memcached)
if err != nil {
	if errors.IsNotFound(err) {
		// Request object not found, could have been deleted after reconcile request.
		// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
		// Return and don't requeue
		reqLogger.Info("Memcached resource not found. Ignoring since object must be deleted")
		return reconcile.Result{}, nil
	}
...
```

> For a guide on Reconcilers, Clients, and interacting with resource Events, see the [Client API doc][doc_client_api].

Reviewer: ```suggestion
For a guide on Reconcilers, Clients, and interacting with resource Events, see the [Client API doc][doc_client_api] and the [controller-runtime documentation over reconcile][controller-runtime-reconcile-godoc]. 
```

At the bottom. 

> [controller-runtime-reconcile-godoc]: https://godoc.org/sigs.k8s.io/controller-runtime/pkg/reconcile

Author: done

```

#### 3. Without an error and no need to requeue the request:

In some situations, such as when the primary resource has been deleted, there is no need to
requeue the request for another attempt

**Usage**
```Go
return reconcile.Result{}, nil
```

**Example:**

In the example below a `reconcile.Result{}, nil` is used because the Memcached resource was not found, and no further processing is required.  
```Go
// Fetch the Memcached instance
memcached := &cachev1alpha1.Memcached{}
err := r.client.Get(context.TODO(), request.NamespacedName, memcached)
if err != nil {
	if errors.IsNotFound(err) {
		// Request object not found, could have been deleted after reconcile request.
		// Owned objects are automatically garbage collected. For additional cleanup logic use finalizers.
		// Return and don't requeue
		reqLogger.Info("Memcached resource not found. Ignoring since object must be deleted")
		return reconcile.Result{}, nil
	}
...
```

For a guide on Reconcilers, Clients, and interacting with resource Events, see the [Client API doc][doc_client_api] and the [controller-runtime documentation over reconcile][controller-runtime-reconcile-godoc].

> 

Reviewer: Broken link. It is not passing in the tests.
> Running markdown link checker against changed markdown files
> Found broken reference (controller-runtime-reconcile-godoc -> ) at ./doc/user-guide.md:334
> Makefile:206: recipe for target 'test-markdown' failed
> make: *** [test-markdown] Error 1
> The command "make test-markdown" exited with 2.

Reviewer: Solved

**Note:** Returning `Result` with `RequeueAfter` set is how you can periodically reconcile a CR.

> #### Reconcile Result Use Cases

Reviewer: ```suggestion
**Example of Reconcile Result Use Cases**
```

Author: Hi @camilamacedo86  Thanks for the suggestion.  I had a crazy week last week and wasn't able to look at this but will look at this soon.

Author: Hi @camilamacedo86 I just got an email that said the following.  Given what the email says below, should I hold off on trying to update this section of the user guide until docs are fully moved?

We are working to build a documentation website that will replace the in-tree github rendered docs.

The initial bootstrapping is in place, docs are temporarily viewable at operator-sdk.netlify.com. Additionally, all PRs now include a netlify preview build.
MERGED: https://github.com/operator-framework/operator-sdk/pull/2667
MERGED: https://github.com/operator-framework/operator-sdk/pull/2679
existing documentation remains in place under doc/
a copy of the documentation is in website/content/en/docs/
[Important] Please do not merge new docs changes to doc/ unless it is important. If changes must be made, please make them in both doc/ and website/content/en/docs/. We have an additional check in place to help make sure this works.

Reviewer: Yes, if we do changes now we need to ensure that the same will be applied in the `website/content/en/docs/` structure as well. 

So, if you would like to move forward with, we need review and be ok with the changes then, after that, do the rebase and ensure that the website structure will have the same content.

Author: Hi @camilamacedo86.  Your proposed approach sounds good.  I just did a rebase and will start updating the content as you recommended

Author: Hi @camilamacedo86 I have made the revisions to both files following your suggested revisions

[event_filtering]:./user/event-filtering.md
[controller_options]: https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/controller#Options
[controller_godocs]: https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/controller

> controller-runtime-reconcile-godoc]: https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/reconcile#Reconciler

Reviewer: ```suggestion
[controller-runtime-reconcile-godoc]: https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/reconcile#Reconciler
```

Author: Yep... It's Monday :-). I saw that and fixed it.  Thanks!
